File: leet_easy_c++/leet_543/leet_543b.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if(!root) return 0;
        leg_depth(root);
        return diam;
    }
    
    int leg_depth(TreeNode* iter){
        if(!iter) return 0;
        int left = leg_depth(iter->left), right = leg_depth(iter->right);
        diam = max(diam, left + right);
        return 1 + max(left, right);
    }
private:
    int diam; //instantiated to zero;
};

//this solution 6ms (amongst fastest solutions)

File: leet_easy_java/leet_543/leet_543b.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private int diam;
    public int diameterOfBinaryTree(TreeNode root) {
        diam = 0;
        DFS(root);
        return diam;
    }
    public int DFS(TreeNode iter){
        if(iter == null) return 0;
        int _left = DFS(iter.left), _right = DFS(iter.right);
        diam = Math.max(_left+_right, diam);
        return Math.max(_left, _right) + 1;
    }
}